### PR TITLE
chore: separate common helpers and add mix delay/partition chaos test.

### LINKED ISSUE #1048: Add e2e test for mix of network delay and partition
After #1031, add network delay will also run iptables commands, so we need to test whether they can work together.

### PR DESCRIPTION
**Pull request by Colstuwjx**
Signed-off-by: colstuwjx <colstuwjx@gmail.com>

### What problem does this PR solve?

Try to fix #1048 .

### What is changed and how does it work?
Add the mixed network partition and delay chaos inside test cases.

### Checklist

Tests

- [ ] Unit test
- [x] E2E test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Breaking backward compatibility

Related changes

- [ ] Need to update the documentation

### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: test/e2e/chaos/basic.go [hunk 1]
```diff
@@ -1726,65 +1637,126 @@ selector:
...
 				framework.ExpectNoError(err, "create network chaos error")
 				time.Sleep(5 * time.Second)
-				framework.ExpectEqual(allSlowConnection(), [][]int{{0, 1}, {0, 2}, {0, 3}})
+				framework.ExpectEqual(allSlowConnection(c, networkPeers, ports), [][]int{{0, 1}, {0, 2}, {0, 3}})
+
+				cancel()
+			})
+		})
+
+		ginkgo.Context("[NetworkPartitionWithNetem]", func() {
```
Comment (Colstuwjx) [inline test/e2e/chaos/basic.go:1696]: I'm not sure whether this test case meets the requirements from issue #1048, since it want to test ipset rules works with iptables changes, and I have no idea about to add the corner case with potential conflicts for them. Any suggestions is welcome :)

Comment (YangKeao) [reply test/e2e/chaos/basic.go:1696]: `Netem` related NetworkChaos adds `iptables` rules iff the `.Spec.Target` field is set.

File: test/e2e/chaos/basic.go [hunk 2]
```diff
@@ -1799,6 +1771,100 @@ selector:
 
 })
 
+func makeNetworkDelayChaos(
```
Comment (YangKeao) [inline test/e2e/chaos/basic.go:1774]: Should we add a `toLabelSelectors` for this function? or we would set `.Spec.Target` manually (e.g. [this line](https://github.com/chaos-mesh/chaos-mesh/pull/1069/files#diff-6370badef8e5715e4f72fad6d6488b517cd1cca6f98764aa93209e5d3710e36bR1614))

**Review by YangKeao — COMMENTED** [review]
Thanks for your contribution! :beers:

Rest LGTM

File: test/e2e/chaos/basic.go (hunk 2 above)
Comment (Colstuwjx) [reply test/e2e/chaos/basic.go:1774]: OK, I agree.

**Comment by Colstuwjx** [comment]
@YangKeao Thanks for your review, PTAL again, thanks!

**Comment by Colstuwjx** [comment]
I'm not sure why the test case rans with different result than I expected:

```

[2020-10-19T07:08:09.852Z] [5] • Failure [70.019 seconds]

[2020-10-19T07:08:09.852Z] [5] [Basic]

[2020-10-19T07:08:09.852Z] [5] /home/jenkins/agent/workspace/chaos-mesh-e2e/go/src/github.com/chaos-mesh/chaos-mesh/test/e2e/chaos/basic.go:58

[2020-10-19T07:08:09.852Z] [5]   [NetworkChaos]

[2020-10-19T07:08:09.852Z] [5]   /home/jenkins/agent/workspace/chaos-mesh-e2e/go/src/github.com/chaos-mesh/chaos-mesh/test/e2e/chaos/basic.go:1409

[2020-10-19T07:08:09.852Z] [5]     [Netem]

[2020-10-19T07:08:09.852Z] [5]     /home/jenkins/agent/workspace/chaos-mesh-e2e/go/src/github.com/chaos-mesh/chaos-mesh/test/e2e/chaos/basic.go:1614

[2020-10-19T07:08:09.852Z] [5]       [Schedule] [It]

[2020-10-19T07:08:09.852Z] [5]       /home/jenkins/agent/workspace/chaos-mesh-e2e/go/src/github.com/chaos-mesh/chaos-mesh/test/e2e/chaos/basic.go:1615

[2020-10-19T07:08:09.852Z] [5] 

[2020-10-19T07:08:09.852Z] [5]       Oct 19 07:08:09.438: Expected

[2020-10-19T07:08:09.852Z] [5]           <[][]int | len:1, cap:1>: [[0, 2]]

[2020-10-19T07:08:09.852Z] [5]       to equal

[2020-10-19T07:08:09.852Z] [5]           <[][]int | len:3, cap:3>: [[0, 1], [0, 2], [0, 3]]

[2020-10-19T07:08:09.852Z] [5] 

[2020-10-19T07:08:09.852Z] [5]       /home/jenkins/agent/workspace/chaos-mesh-e2e/go/src/github.com/chaos-mesh/chaos-mesh/test/e2e/chaos/basic.go:1675
```

**Comment by YangKeao** [comment]
Oops. I have run the e2e tests several times in my cluster. This error doesn't occur. However, there is another problem: `allSlowConnection` will throw an exception when the request timeout. However, the timeout is the expected behavior of network partition.

An ugly fix could be replacing `networkPeers, ports` with `networkPeers[1:], ports[1:]` (and the index corresponding) to avoid sending request from/to the blocked peer.

A better solution could be refractoring these two functions into one `allNetworkAbnormalCondition` (or a better name).

File: test/e2e/chaos/basic.go [hunk 3]
```diff
@@ -1640,151 +1621,210 @@ selector:
...
+					}
+				)
+
+				networkPartition := makeNetworkPartitionChaos(
+					ns, "network-chaos-1",
+					map[string]string{"app": "network-peer-0"},
+					map[string]string{"app": "network-peer-1"},
+					v1alpha1.OnePodMode,
+					v1alpha1.OnePodMode,
+					v1alpha1.To,
```
Comment (YangKeao) [inline test/e2e/chaos/basic.go:1790]: It will only inject iptables rules on `network-peer-0`

Comment (Colstuwjx) [reply test/e2e/chaos/basic.go:1820]: Okay, I changed to 2->0, so now peer-2 partitioned with peer-0, and delayed with peer-1.

**Comment by Colstuwjx** [comment]
> A better solution could be refractoring these two functions into one allNetworkAbnormalCondition (or a better name).

I choose another option, add an except list for avoiding running delays on these pairs.

File: test/e2e/chaos/basic.go [hunk 4]
```diff
@@ -1640,151 +1637,224 @@ selector:
...
+					testDelayDuration,
+					testDelaySchedulerSpec,
+				)
+
 				klog.Infof("Injecting delay for 0")
 				err = cli.Create(ctx, networkDelay.DeepCopy())
 				framework.ExpectNoError(err, "create network chaos error")
 				time.Sleep(5 * time.Second)
-				framework.ExpectEqual(allSlowConnection(), [][]int{{0, 1}, {0, 2}, {0, 3}})
+				framework.ExpectEqual(allSlowConnection(c, networkPeers, ports, noExcepts), [][]int{{0, 1}, {0, 2}, {0, 3}})
```
Comment (Colstuwjx) [inline test/e2e/chaos/basic.go:1695]: This case still return result `[][]int{{0, 1}}`, and after check the logs:

```

[2020-10-20T07:38:17.102Z] [5] Oct 20 07:38:16.833: INFO: network-peer-2-6d4d849666-gsvbl  chaos-mesh-worker2  Running         [{Initialized True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 07:37:21 +0000 UTC  } {Ready True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 07:37:22 +0000 UTC  } {ContainersReady True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 07:37:22 +0000 UTC  } {PodScheduled True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 07:37:21 +0000 UTC  }]

[2020-10-20T07:38:17.102Z] [5] Oct 20 07:38:16.833: INFO: network-peer-1-66b96cb95b-njm9z  chaos-mesh-worker3  Running         [{Initialized True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 07:37:16 +0000 UTC  } {Ready True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 07:37:18 +0000 UTC  } {ContainersReady True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 07:37:18 +0000 UTC  } {PodScheduled True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 07:37:16 +0000 UTC  }]

```

I found both of `network-peer-0`, `network-peer-2`, `network-peer-3` are deployed on chaos-mesh-worker2, and `network-peer-1` deployed on another node `chaos-mesh-worker3`. I'm not sure if it's the root cause...

File: test/e2e/chaos/basic.go [hunk 5]
```diff
@@ -1640,151 +1637,224 @@ selector:
...
+					testDelayTcParam,
+					testDelayDuration,
+					testDelaySchedulerSpec,
+				)
+
+				klog.Infof("Injecting delay for 2 for mix chaos")
+				err = cli.Create(ctx, networkDelay.DeepCopy())
+				framework.ExpectNoError(err, "create network chaos error")
+				time.Sleep(5 * time.Second)
+				framework.ExpectEqual(allSlowConnection(c, networkPeers, ports, exceptPartitionPair), [][]int{{0, 2}, {2, 1}})
```
Comment (Colstuwjx) [inline test/e2e/chaos/basic.go:1847]: It seems that mix chaos would make peer-0 couldn't visit peer-2, and our spec is:

```
					ns, "network-chaos-1",
					map[string]string{"app": "network-peer-2"},
					map[string]string{"app": "network-peer-0"},
					v1alpha1.OnePodMode,
					v1alpha1.OnePodMode,
					v1alpha1.To
```

it's direction is `network-peer-2` **To** `network-peer-0`. The expected result should be `peer-2` could NOT visit `peer-0`, but `peer-0` could visit `peer-2`, right? @YangKeao

File: test/e2e/chaos/basic.go (hunk 4 above)
Comment (Colstuwjx) [reply test/e2e/chaos/basic.go:1695]: At the latest build, I see the logs below:

```
[2020-10-20T08:53:14.750Z] [5] I1020 08:53:14.629320      23 basic.go:1691] Injecting delay for 0
[2020-10-20T08:53:20.007Z] [5] I1020 08:53:19.643029      23 basic.go:1967] testing delay from network-peer-0-68599f7689-dm4p5 to network-peer-1-54bb55d857-cxdwt
[2020-10-20T08:53:20.007Z] [5] I1020 08:53:19.643083      23 basic.go:2068] sending request to localhost:37948 with body: {"targetIP":"10.244.3.5"}
[2020-10-20T08:53:20.007Z] [5] I1020 08:53:19.645379      23 basic.go:1973] delay from 0 to 1: 513566
[2020-10-20T08:53:20.007Z] [5] I1020 08:53:19.645395      23 basic.go:1967] testing delay from network-peer-0-68599f7689-dm4p5 to network-peer-2-65c886b4b7-94756
[2020-10-20T08:53:20.008Z] [5] I1020 08:53:19.645405      23 basic.go:2068] sending request to localhost:37948 with body: {"targetIP":"10.244.1.9"}
[2020-10-20T08:53:20.008Z] [5] I1020 08:53:19.647108      23 basic.go:1973] delay from 0 to 2: 308449
[2020-10-20T08:53:20.008Z] [5] I1020 08:53:19.647133      23 basic.go:1967] testing delay from network-peer-0-68599f7689-dm4p5 to network-peer-3-5d4b896675-kzmkg
[2020-10-20T08:53:20.008Z] [5] I1020 08:53:19.647144      23 basic.go:2068] sending request to localhost:37948 with body: {"targetIP":"10.244.3.7"}
[2020-10-20T08:53:20.008Z] [5] I1020 08:53:19.848924      23 basic.go:1973] delay from 0 to 3: 200461654
[2020-10-20T08:53:20.008Z] [5] I1020 08:53:19.848947      23 basic.go:1967] testing delay from network-peer-1-54bb55d857-cxdwt to network-peer-2-65c886b4b7-94756
[2020-10-20T08:53:20.008Z] [5] I1020 08:53:19.848957      23 basic.go:2068] sending request to localhost:33231 with body: {"targetIP":"10.244.1.9"}
[2020-10-20T08:53:20.008Z] [5] I1020 08:53:19.850735      23 basic.go:1973] delay from 1 to 2: 355391
[2020-10-20T08:53:20.008Z] [5] I1020 08:53:19.850754      23 basic.go:1967] testing delay from network-peer-1-54bb55d857-cxdwt to network-peer-3-5d4b896675-kzmkg
[2020-10-20T08:53:20.008Z] [5] I1020 08:53:19.850764      23 basic.go:2068] sending request to localhost:33231 with body: {"targetIP":"10.244.3.7"}
[2020-10-20T08:53:20.008Z] [5] I1020 08:53:19.852280      23 basic.go:1973] delay from 1 to 3: 267228
[2020-10-20T08:53:20.008Z] [5] I1020 08:53:19.852297      23 basic.go:1967] testing delay from network-peer-2-65c886b4b7-94756 to network-peer-3-5d4b896675-kzmkg
[2020-10-20T08:53:20.008Z] [5] I1020 08:53:19.852307      23 basic.go:2068] sending request to localhost:36099 with body: {"targetIP":"10.244.3.7"}
[2020-10-20T08:53:20.008Z] [5] I1020 08:53:19.853873      23 basic.go:1973] delay from 2 to 3: 299827
[2020-10-20T08:53:20.008Z] [5] Oct 20 08:53:19.853: FAIL: Expected
[2020-10-20T08:53:20.008Z] [5]     <[][]int | len:1, cap:1>: [[0, 3]]
[2020-10-20T08:53:20.008Z] [5] to equal
[2020-10-20T08:53:20.008Z] [5]     <[][]int | len:3, cap:3>: [[0, 1], [0, 2], [0, 3]]
```

```
[2020-10-20T08:55:18.941Z] [2] Oct 20 08:55:18.237: INFO: network-peer-0-68599f7689-glgrr  chaos-mesh-worker2  Running         [{Initialized True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 08:53:57 +0000 UTC  } {Ready True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 08:53:59 +0000 UTC  } {ContainersReady True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 08:53:59 +0000 UTC  } {PodScheduled True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 08:53:57 +0000 UTC  }]
[2020-10-20T08:55:18.941Z] [2] Oct 20 08:55:18.237: INFO: network-peer-1-54bb55d857-z7fm5  chaos-mesh-worker3  Running         [{Initialized True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 08:54:02 +0000 UTC  } {Ready True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 08:54:04 +0000 UTC  } {ContainersReady True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 08:54:04 +0000 UTC  } {PodScheduled True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 08:54:02 +0000 UTC  }]
[2020-10-20T08:55:18.941Z] [2] Oct 20 08:55:18.237: INFO: network-peer-2-65c886b4b7-rkwdn  chaos-mesh-worker3  Running         [{Initialized True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 08:54:07 +0000 UTC  } {Ready True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 08:54:09 +0000 UTC  } {ContainersReady True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 08:54:09 +0000 UTC  } {PodScheduled True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 08:54:07 +0000 UTC  }]
[2020-10-20T08:55:18.941Z] [2] Oct 20 08:55:18.237: INFO: network-peer-3-5d4b896675-p2vh2  chaos-mesh-worker2  Running         [{Initialized True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 08:54:13 +0000 UTC  } {Ready True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 08:54:15 +0000 UTC  } {ContainersReady True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 08:54:15 +0000 UTC  } {PodScheduled True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 08:54:13 +0000 UTC  }]
```

Okay, at this time, it's peer-3...

Comment (YangKeao) [reply test/e2e/chaos/basic.go:1695]: We judged whether the target is set by comparing `.Spec.Target` with `nil`. If `.Spec.Target` is not nil but it contains no selector, it will `List` the pods under the namespace.

And you select one pod from the pods, so the result seems to be random.

To test `netem` without filter, please remove the `.Spec.Target` field (or set it to nil).

Comment (Colstuwjx) [reply test/e2e/chaos/basic.go:1695]: OK, many thanks!

**Comment by STRRL** [comment]
Hi @Colstuwjx , sorry for the late response. 

It seems here is a problem about `couldConnect()`([link](https://github.com/Colstuwjx/chaos-mesh/blob/chore%2Fmix-net-delay-partition-chaos-test/test/e2e/chaos/basic.go#L2127)). It tries to receive UDP packet IMMEDIATELY after sending UDP packet, it will not tolerate any network delay. So the e2e  testcase `NetworkPartitionWithNetem` always goes failed, `NetworkPartition ` is also unstable. 

I think here is no way to ACK some UDP packets received. So using `time.Sleep()` for a "reasonable" time between `sendUDPPacket()` and `recvUDPPacket()` will resolve this. It's not clean, but it will work on most situation. 

Any other idea about this? @YangKeao

**Comment by YangKeao** [comment]
> using `time.Sleep()` for a "reasonable" time(1 or 2 seconds) 

Sleep is OK. In the e2e test, there are already a lot of unreasonable sleep :laughing: .

**Comment by Colstuwjx** [comment]
@STRRL Thank you for your review! When I developing this PR, I was wondering why the previous cases also broken, and you gave a likely answer for it. It helps me a lot!

> I think here is no way to ACK some UDP packets received. So using time.Sleep() for a "reasonable" time(1 or 2 seconds) between sendUDPPacket() and recvUDPPacket() will resolve this. It's not clean, but it will work on most situation.

I'll give a try. BTW, is this issue introduced by my changes, or it has been exists for a long time, I still doubt about this.

**Comment by STRRL** [comment]
It seems be there for a long time. 🙈

**Comment by Colstuwjx** [comment]
It seems that the error is not related to this PR:

```
[2020-11-06T08:53:49.951Z] [5] Nov  6 08:53:49.636: FAIL: pod failure recover failed

[2020-11-06T08:53:49.951Z] [5] Unexpected error:

[2020-11-06T08:53:49.951Z] [5]     <*errors.errorString | 0xc0003c92b0>: {

[2020-11-06T08:53:49.951Z] [5]         s: "timed out waiting for the condition",

[2020-11-06T08:53:49.951Z] [5]     }

[2020-11-06T08:53:49.951Z] [5]     timed out waiting for the condition

[2020-11-06T08:53:49.951Z] [5] occurred

[2020-11-06T08:53:49.951Z] [5] [AfterEach] [Basic]

[2020-11-06T08:53:49.951Z] [5]   /go/pkg/mod/k8s.io/kubernetes@v1.17.2/test/e2e/framework/framework.go:152

[2020-11-06T08:53:49.951Z] [5] STEP: Collecting events from namespace "chaos-mesh-3981".

[2020-11-06T08:53:49.951Z] [5] STEP: Found 12 events.
```

It comes from [this line](https://github.com/chaos-mesh/chaos-mesh/blob/master/test/e2e/chaos/basic.go#L173), which is part of PodFailureChaos experiment. In other word, this part may be unstable.

**Comment by Colstuwjx** [comment]
@YangKeao @STRRL I think this PR is ready to merge :)

**Comment by STRRL** [comment]
> It seems that the error is not related to this PR:
> 
> ```
> [2020-11-06T08:53:49.951Z] [5] Nov  6 08:53:49.636: FAIL: pod failure recover failed
> 
> [2020-11-06T08:53:49.951Z] [5] Unexpected error:
> 
> [2020-11-06T08:53:49.951Z] [5]     <*errors.errorString | 0xc0003c92b0>: {
> 
> [2020-11-06T08:53:49.951Z] [5]         s: "timed out waiting for the condition",
> 
> [2020-11-06T08:53:49.951Z] [5]     }
> 
> [2020-11-06T08:53:49.951Z] [5]     timed out waiting for the condition
> 
> [2020-11-06T08:53:49.951Z] [5] occurred
> 
> [2020-11-06T08:53:49.951Z] [5] [AfterEach] [Basic]
> 
> [2020-11-06T08:53:49.951Z] [5]   /go/pkg/mod/k8s.io/kubernetes@v1.17.2/test/e2e/framework/framework.go:152
> 
> [2020-11-06T08:53:49.951Z] [5] STEP: Collecting events from namespace "chaos-mesh-3981".
> 
> [2020-11-06T08:53:49.951Z] [5] STEP: Found 12 events.
> ```
> 
> It comes from [this line](https://github.com/chaos-mesh/chaos-mesh/blob/master/test/e2e/chaos/basic.go#L173), which is part of PodFailureChaos experiment. In other word, this part may be unstable.

That's wired. Could you re-produce this problem and provide some logs about controller-manager?

**Comment by Colstuwjx** [comment]
> That's wired. Could you re-produce this problem and provide some logs about controller-manager?

It's from CI jenkins job log, and I've tried some times on my local kind cluster, it didn't happen again.

**Comment by YangKeao** [comment]
`PodFailure` e2e test seems not stable (in our CI environment). The reason could be the poor network condition or lack of computing resource @STRRL .

File: test/e2e/chaos/basic.go [hunk 6]
```diff
@@ -49,6 +49,32 @@ import (
...
+type exceptMap struct {
+	// Source is the source pod labelSelector of chaos
+	Source map[string]string
+	// Target is the target pod labelSelector of chaos
+	Target map[string]string
+}
+
+type exceptMaps []*exceptMap
+
+func (e exceptMaps) isExcept(source, target map[string]string) bool {
```
Comment (STRRL) [inline test/e2e/chaos/basic.go:61]: Is there any way to avoid `exceptMaps`? Like combinating `allSlowConnection()` and `allBlockedConnection()` as one function.

It seems `exceptMaps` wants to avoiding throw error in `allSlowConnection()`.

**Review by STRRL — CHANGES_REQUESTED** [review]
rest LGTM

File: test/e2e/chaos/basic.go (hunk 6 above)
Comment (Colstuwjx) [reply test/e2e/chaos/basic.go:61]: Mixing `allSlowConnection` and `allBlockedConnection` into one function sounds not a good idea, so what's this combined function meaning? It's ambiguous!

Adding `excepMaps` would skip some cases while calling `allSlowConnection`, it's not just ONLY suitable to this PR, but also for the other potential issues, and I think this design would be more commonly to use.

File: test/e2e/chaos/basic.go [hunk 7]
```diff
@@ -1640,151 +1647,236 @@ selector:
...
+
+				err = cli.Delete(ctx, networkPartition.DeepCopy())
+				framework.ExpectNoError(err, "delete network chaos error")
+				time.Sleep(5 * time.Second)
+				framework.ExpectEqual(len(allBlockedConnection(c, networkPeers, ports)), 0)
+
+				err = cli.Delete(ctx, networkDelay.DeepCopy())
+				framework.ExpectNoError(err, "delete network chaos error")
+				time.Sleep(5 * time.Second)
+				framework.ExpectEqual(len(allSlowConnection(c, networkPeers, ports, noExcepts)), 0)
```
[the remark this answers is not in the document]
Comment (Colstuwjx) [reply test/e2e/chaos/basic.go:1879]: Okay.

File: test/e2e/chaos/basic.go (hunk 6 above)
Comment (STRRL) [reply test/e2e/chaos/basic.go:61]: How about making a method that return a thing like `map[EnumOfNetworkStates][][]int`?

I just do NOT want to skip making asserting on something. 
And more, I think we could improve these functions which could helper us to detect network status.

We pretend that there are three states for network: good, slow and blocked, then
we using these methods during assertion in e2e test. 

Currently `allSlowConnection` is not enough expressively with both network partition and delay. It assumes that network communications between each peer are all good, except few connections is slow. Using `allSlowConnection` with a `expectMaps` just skipping connections which setted with network-partition, then using `allBlockedConnection` to assert the network-partition exists is acceptable. We do not need do any changes if we don't have another brilliant ideas.

Comment (Colstuwjx) [reply test/e2e/chaos/basic.go:61]: Got it, I'll try :)

File: test/e2e/chaos/basic.go [hunk 8]
```diff
@@ -1640,151 +1660,262 @@ selector:
...
+
+				klog.Infof("Injecting delay for 0 with more complicate netem")
 				err = cli.Create(ctx, complicateNetem.DeepCopy())
 				framework.ExpectNoError(err, "create network chaos error")
 				time.Sleep(5 * time.Second)
-				framework.ExpectEqual(allSlowConnection(), [][]int{{0, 1}, {0, 2}, {0, 3}})
+
+				result = probeNetworkCondition(c, networkPeers, ports)
+
+				framework.ExpectEqual(len(result[networkConditionBlocked]), 0)
```
Comment (Colstuwjx) [inline test/e2e/chaos/basic.go:1820]: @STRRL After merged into `probeNetworkCondition`, I found that the `couldConnect` network test  would broken, since UDP is an unreliable protocol under the bad network.

From now, I workaround this, print out the log and still return true, so it's still 0 at here.

Comment (STRRL) [reply test/e2e/chaos/basic.go:1820]: You mean `time.Sleep(2 * time.Second)` between `sendUDPPacket` and `recvUDPPacket`?

Comment (Colstuwjx) [reply test/e2e/chaos/basic.go:1820]: Nope. In the previous, we didn't do `couldConnect` test at here, we called `allSlowConnection` and it ONLY did `testDelay`, as we all known, `testDelay` would send HTTP post request, which is a TCP-based request, and `couldConnect` is an UDP request.

At here, we're testing the complicate netem delay case, and it would make each network request may lost packet, in other words, it may also make the `couldConnect` return bad result. After we make the two functions merged into `probeNetworkCondition`, I met this issue! `couldConnect` may receive empty data(lost since it's UDP request), and I workaround it, see [this line](https://github.com/chaos-mesh/chaos-mesh/pull/1069/files#diff-6370badef8e5715e4f72fad6d6488b517cd1cca6f98764aa93209e5d3710e36bR2188).

Comment (STRRL) [reply test/e2e/chaos/basic.go:1820]: I guess that this issue ONLY happens on `testDelayTcParamEvenMoreComplicate`, does it?  
It makes delay, loss, duplicate and corrupt at same time. Honestly I have no idea for making assertion stable under this situation. 🙈

Comment (Colstuwjx) [reply test/e2e/chaos/basic.go:1820]: > I guess that this issue ONLY happens on testDelayTcParamEvenMoreComplicate, does it?
Yes, so I choose workaround :(

**Comment by Colstuwjx** [comment]
@STRRL @YangKeao Everything cleared, I think this PR is ready to merge :)

**Review by STRRL — APPROVED** [review]
LGTM

**Review by YangKeao — APPROVED** [review]
LGTM
